File: backend/app/schemas/messaging.py
from pydantic import BaseModel, Field
from typing import Optional
from uuid import UUID
from datetime import datetime


# ============================================================
# Conversation Schemas
# ============================================================


class ConversationCreate(BaseModel):
    participant_ids: list[UUID]
    title: Optional[str] = Field(None, max_length=255)
    is_group: bool = False


class ConversationUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    group_photo_url: Optional[str] = None
    chat_theme: Optional[str] = None
    is_pinned: Optional[bool] = None
    is_archived: Optional[bool] = None


class ConversationMemberResponse(BaseModel):
    id: UUID
    user_id: UUID
    role: str
    is_muted: bool
    is_notifications_paused: bool
    is_pinned: bool
    is_archived: bool
    last_read_at: Optional[datetime]
    joined_at: datetime
    is_left: bool
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    is_online: bool = False
    last_seen_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ConversationResponse(BaseModel):
    id: UUID
    title: Optional[str]
    is_group: bool
    group_photo_url: Optional[str]
    created_by_id: Optional[UUID]
    last_message_id: Optional[UUID]
    last_message_at: Optional[datetime]
    is_archived: bool
    is_pinned: bool
    chat_theme: str
    created_at: datetime
    updated_at: datetime
    members: list[ConversationMemberResponse] = []
    unread_count: int = 0
    last_message_content: Optional[str] = None
    last_message_sender_name: Optional[str] = None

    model_config = {"from_attributes": True}


# ============================================================
# Message Schemas
# ============================================================


class MessageCreate(BaseModel):
    content: Optional[str] = None
    message_type: str = Field(default="text", pattern=r"^(text|image|video|audio|file|voice|gif|sticker|system)$")
    media_url: Optional[str] = None
    media_id: Optional[UUID] = None
    thumbnail_url: Optional[str] = None
    file_name: Optional[str] = None
    file_size: Optional[int] = None
    mime_type: Optional[str] = None
    duration: Optional[float] = None
    reply_to_id: Optional[UUID] = None
    metadata_json: Optional[str] = None


class MessageUpdate(BaseModel):
    content: Optional[str] = None
    is_edited: bool = True


class MessageReactionCreate(BaseModel):
    emoji: str = Field(..., max_length=10)


class MessageResponse(BaseModel):
    id: UUID
    conversation_id: UUID
    sender_id: Optional[UUID]
    content: Optional[str]
    message_type: str
    media_url: Optional[str]
    media_id: Optional[UUID]
    thumbnail_url: Optional[str]
    file_name: Optional[str]
    file_size: Optional[int]
    mime_type: Optional[str]
    duration: Optional[float]
    reply_to_id: Optional[UUID]
    is_edited: bool
    is_deleted: bool
    is_unsent: bool
    reactions_count: int
    reply_count: int
    is_forwarded: bool
    forwarded_from_id: Optional[UUID]
    metadata_json: Optional[str]
    created_at: datetime
    updated_at: datetime
    sender_name: Optional[str] = None
    sender_avatar: Optional[str] = None
    reactions: list["MessageReactionResponse"] = []
    read_by: list["MessageReadResponse"] = []
    reply_to_preview: Optional["MessageResponse"] = None

    model_config = {"from_attributes": True}


class MessageReactionResponse(BaseModel):
    id: UUID
    message_id: UUID
    user_id: UUID
    emoji: str
    created_at: datetime
    username: Optional[str] = None

    model_config = {"from_attributes": True}


class MessageReadResponse(BaseModel):
    id: UUID
    message_id: UUID
    user_id: UUID
    created_at: datetime
    username: Optional[str] = None

    model_config = {"from_attributes": True}


# ============================================================
# Typing / Online Status Schemas
# ============================================================


class TypingIndicatorCreate(BaseModel):
    is_typing: bool = True


class TypingIndicatorResponse(BaseModel):
    user_id: UUID
    conversation_id: UUID
    is_typing: bool
    username: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class OnlineStatusResponse(BaseModel):
    user_id: UUID
    is_online: bool
    last_seen_at: Optional[datetime]
    status_text: Optional[str]

    model_config = {"from_attributes": True}


class OnlineStatusUpdate(BaseModel):
    is_online: bool
    status_text: Optional[str] = None


# ============================================================
# Chat Theme Schemas
# ============================================================


class ChatThemeResponse(BaseModel):
    theme: str
    primary_color: Optional[str] = None
    background_color: Optional[str] = None


# ============================================================
# Pinned / Archived Chat Schemas
# ============================================================


class ChatPinUpdate(BaseModel):
    is_pinned: bool


class ChatArchiveUpdate(BaseModel):
    is_archived: bool


class ChatMuteUpdate(BaseModel):
    is_muted: bool


# ============================================================
# Search Schemas
# ============================================================


class MessageSearchResult(BaseModel):
    messages: list[MessageResponse]
    total_count: int


# ============================================================
# Forward Schemas
# ============================================================


class MessageForward(BaseModel):
    conversation_ids: list[UUID]


# ============================================================
# Read Receipt Schemas
# ============================================================


class MarkAsRead(BaseModel):
    message_id: UUID

File: frontend/src/types/messaging.ts
export interface ConversationMember {
  id: string;
  user_id: string;
  role: "admin" | "member";
  is_muted: boolean;
  is_notifications_paused: boolean;
  is_pinned: boolean;
  is_archived: boolean;
  last_read_at: string | null;
  joined_at: string;
  is_left: boolean;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  is_online: boolean;
  last_seen_at: string | null;
}

export interface Conversation {
  id: string;
  title: string | null;
  is_group: boolean;
  group_photo_url: string | null;
  created_by_id: string | null;
  last_message_id: string | null;
  last_message_at: string | null;
  is_archived: boolean;
  is_pinned: boolean;
  chat_theme: string;
  created_at: string;
  updated_at: string;
  members: ConversationMember[];
  unread_count: number;
  last_message_content: string | null;
  last_message_sender_name: string | null;
}

export interface Message {
  id: string;
  conversation_id: string;
  sender_id: string | null;
  content: string | null;
  message_type: "text" | "image" | "video" | "audio" | "file" | "voice" | "gif" | "sticker" | "system";
  media_url: string | null;
  media_id: string | null;
  thumbnail_url: string | null;
  file_name: string | null;
  file_size: number | null;
  mime_type: string | null;
  duration: number | null;
  reply_to_id: string | null;
  is_edited: boolean;
  is_deleted: boolean;
  is_unsent: boolean;
  reactions_count: number;
  reply_count: number;
  is_forwarded: boolean;
  forwarded_from_id: string | null;
  metadata_json: string | null;
  created_at: string;
  updated_at: string;
  sender_name: string | null;
  sender_avatar: string | null;
  reactions: MessageReaction[];
  read_by: MessageRead[];
  reply_to_preview?: Message | null;
}

export interface MessageReaction {
  id: string;
  message_id: string;
  user_id: string;
  emoji: string;
  created_at: string;
  username: string | null;
}

export interface MessageRead {
  id: string;
  message_id: string;
  user_id: string;
  created_at: string;
  username: string | null;
}

export interface TypingIndicator {
  user_id: string;
  conversation_id: string;
  is_typing: boolean;
  username: string | null;
  created_at: string;
}

export interface OnlineStatus {
  user_id: string;
  is_online: boolean;
  last_seen_at: string | null;
  status_text: string | null;
}

export interface ConversationCreate {
  participant_ids: string[];
  title?: string;
  is_group?: boolean;
}

export interface ConversationUpdate {
  title?: string;
  group_photo_url?: string;
  chat_theme?: string;
  is_pinned?: boolean;
  is_archived?: boolean;
}

export interface MessageCreate {
  content?: string;
  message_type?: string;
  media_url?: string;
  media_id?: string;
  thumbnail_url?: string;
  file_name?: string;
  file_size?: number;
  mime_type?: string;
  duration?: number;
  reply_to_id?: string;
  metadata_json?: string;
}

export interface MessageUpdate {
  content?: string;
  is_edited?: boolean;
}

export interface MessageSearchResult {
  messages: Message[];
  total_count: number;
}

export interface WebSocketMessage {
  type: string;
  conversation_id?: string;
  message?: Message;
  user_id?: string;
  username?: string;
  is_typing?: boolean;
  is_online?: boolean;
  message_id?: string;
  emoji?: string;
  content?: string;
  notification?: {
    id: string;
    type: string;
    entity_type: string;
    entity_id: string;
    content: string | null;
    created_at: string;
    actor?: {
      id: string;
      full_name: string | null;
      username: string | null;
      avatar_url: string | null;
    };
  };
}
